### PR TITLE
Abandoned - replaced with 2959 - Add support for OCS deployment via OCS-CI for IBM Power platform

### PR DESCRIPTION
**Pull request by gitsridhar**
In this commit, I have added support for deploying OCS using OCS-CI. 
The supported storage cluster yaml file is included in the package.
This config yaml file is also included in the package.
A new plugin ibm_deployment.yaml is included in the package. The corresponding deployer is marked as "powervs" and type as "upi".

### CODE REVIEW

**Comment by gitsridhar** [comment]
This is the conf.yaml file related to this work:
RUN:
  username: 'kubeadmin'
  password_location: 'auth/kubeadmin-password'
  log_dir: "/tmp"
  kubeconfig_location: 'auth/kubeconfig' # relative from cluster_dir
# bin_dir: './bin'
  bin_dir: '/usr/local/bin'

DEPLOYMENT:
  local_storage: True
  force_download_installer: False
  force_download_client: False
  ocs_csv_channel: "stable-4.6"
  default_latest_tag: "latest-stable-4.6.0"
ENV_DATA:
  cluster_name: null.  # this parameter is passed in while invoking run-ci
  storage_cluster_name: 'ocs-storagecluster'
  storage_device_sets_name: "ocs-deviceset"
  cluster_namespace: 'openshift-storage'
  ocs_version: '4.6'
  skip_ocp_deployment: true
  skip_ocs_deployment: false
  platform: 'powervs'
  disk_pattern: 'sdn'
  deployment_type: 'upi'

REPORTING:
  email:
    address: "ocs-ci@redhat.com"
    smtp_server: "localhost"
  polarion:
    project_id: "OpenShiftContainerStorage"
  us_ds: 'DS'
  ocp_must_gather_image: "quay.io/rhceph-dev/ocs-must-gather"
  ocs_must_gather_image: "quay.io/rhceph-dev/ocs-must-gather"
  default_ocs_must_gather_latest_tag: 'latest-4.6'
  gather_on_deploy_failure: true

File: ocs_ci/deployment/ibm_deployment.py [hunk 1]
```diff
@@ -0,0 +1,1021 @@
+"""
```
Comment (ebenahar) [inline ocs_ci/deployment/ibm_deployment.py:1]: please call the file `ibm` instead of `ibm_deployment.py` to keep consistency with the other files in this dir

File: ocs_ci/deployment/ibm_deployment.py [hunk 2]
```diff
@@ -0,0 +1,1021 @@
...
+This module provides base class for different deployment
+platforms like AWS, VMWare, Baremetal etc.
```
Comment (ebenahar) [inline ocs_ci/deployment/ibm_deployment.py:3]: Should be IBM only

File: ocs_ci/deployment/ibm_deployment.py [hunk 3]
```diff
@@ -0,0 +1,1021 @@
...
+                )
+                update_ntp_compute_nodes()
+                assert ceph_health_check(
+                    namespace=self.namespace, tries=60, delay=10
+                )
+
+        # patch gp2/thin storage class as 'non-default'
+        self.patch_default_sc_to_non_default()
+
+        # Modify Noobaa endpoint auto scale values according to the cluster specs
```
Comment (ebenahar) [inline ocs_ci/deployment/ibm_deployment.py:683]: I think that there are some core duplicity with ocs_ci/deployment/deployment content, like the section here

File: ocs_ci/deployment/ibm_deployment.py (hunk 1 above)
Comment (gitsridhar) [reply ocs_ci/deployment/ibm_deployment.py:1]: I will make this change along with others in this review and update pull request.

**Review by vasukulkarni — CHANGES_REQUESTED** [review]

File: ocs_ci/deployment/ibm_deployment.py [hunk 4]
```diff
@@ -0,0 +1,1024 @@
...
+
+    def post_ocp_deploy(self):
+        """
+        Function does post OCP deployment stuff we need to do.
+        """
+        set_selinux_permissions()
+        set_registry_to_managed_state()
+        add_stage_cert()
+
+    def label_and_taint_nodes(self):
```
[the remark this answers is not in the document]
Comment (clacroix12) [reply ocs_ci/deployment/ibm_deployment.py:160]: We definitely don't want to be duplicating what is already established within the base class. That defeats the purpose of inheriting from that class in the first place and makes it much more difficult to determine what is specifically done for the IBM deployment vs. other deployments. It appears that the entire deployment module has been copy/pasted here and then potentially changed for IBM specific functionality. It would be better to start fresh with a new class that inherits from `Deployment` and only add functionality / override functions as necessary.

**Review by clacroix12 — CHANGES_REQUESTED** [review]

File: ocs_ci/deployment/ibm_deployment.py (hunk 1 above)
Comment (gitsridhar) [reply ocs_ci/deployment/ibm_deployment.py]: Completed. I moved ibm_deployment.py to ibm.py

File: ocs_ci/deployment/ibm_deployment.py (hunk 3 above)
Comment (gitsridhar) [reply ocs_ci/deployment/ibm_deployment.py]: I removed redundant code. ibm.py has only whats used in it.

File: ocs_ci/deployment/ibm_deployment.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (gitsridhar) [reply ocs_ci/deployment/ibm_deployment.py]: I trimmed ibm_deployment.py, renamed to ibm.py and removed redundant code.

**Comment by gitsridhar** [comment]
I addressed all the comments. I also implemented 'teardown' and is under test now. I will update this PR comments for the final review.

File: ocs_ci/deployment/ibm.py [hunk 5]
```diff
@@ -0,0 +1,548 @@
...
+        # STEP 7 & 9
+        nodes = ocp.OCP(kind='node').get().get('items', [])
+        worker_nodes = [
+            node for node in nodes if "cluster.ocs.openshift.io/openshift-storage"
+            in node['metadata']['labels']
+        ]
+        worker_node_names = [node['metadata']['name'] for node in worker_nodes]
+        logger.warning(worker_node_names)
+
+        # for worker_node in worker_node_names:
```
Comment (vasukulkarni) [inline ocs_ci/deployment/ibm.py:252]: Please remove the unused commented code

File: ocs_ci/deployment/ibm.py [hunk 6]
```diff
@@ -0,0 +1,548 @@
...
+        # STEP 8
+        for sc in scandlvlist:
+            lvdelete = 'oc delete localvolume -n local-storage --wait=true ' + sc[0]
+            logger.warning(lvdelete)
+            # run_cmd(lvdelete)
+            pvdelete = 'oc delete pv -l \
+                storage.openshift.com/local-volume-owner-name=' + sc[0] \
+                + ' --wait --timeout=5m'
+            logger.warning(pvdelete)
+            # run_cmd(pvdelete)
```
Comment (vasukulkarni) [inline ocs_ci/deployment/ibm.py:273]: most of the run_cmd are commented out, are you not deleting them?

File: ocs_ci/deployment/ibm.py [hunk 7]
```diff
@@ -0,0 +1,548 @@
...
+        noobaas.noobaa.io ocsinitializations.ocs.openshift.io \
+        storageclusterinitializations.ocs.openshift.io \
+        storageclusters.ocs.openshift.io  --wait=true --timeout=5m'
+        logger.warning(crds)
+        # run_cmd(crds)
+
+    def deploy_ocs(self):
+        """
+        Handle OCS deployment, since OCS deployment steps are common to any
+        platform, implementing OCS deployment here in base class.
```
Comment (clacroix12) [inline ocs_ci/deployment/ibm.py:315]: Like this docstring suggests, OCS deployment should be common regardless of platform so the deployment is implemented in the base class. If there IS something that needs to change for a particular platform, we should have some conditional code in the base class that handles it. We shouldn't overriding the `deploy_ocs` method with mostly the same code.

File: ocs_ci/deployment/ibm.py [hunk 8]
```diff
@@ -0,0 +1,548 @@
...
+    def destroy_cluster(self, log_level="DEBUG"):
+        """
+        Base destroy cluster method, for more platform specific stuff please
+        overload this method in child class.
+
+        Args:
+            log_level (str): log level for installer (default: DEBUG)
+        """
+        if not config.ENV_DATA['skip_ocs_deployment']:
+            self.destroy_ocs()
```
Comment (clacroix12) [inline ocs_ci/deployment/ibm.py:478]: Is this necessary since by destroying OCP we inherently destroy OCS as well?

**Review by clacroix12 — COMMENTED** [review]
Seems like this PR is still a work in progress. I suggest adding `WIP` to the title so reviewers know as such. I think there is still a lot of duplication here that needs to be pulled out. Feel free to ping me (or anyone in @red-hat-storage/ocs-qe-ecosystem) if you have any questions about the class structure/workflow!

File: ocs_ci/deployment/ibm_deployment.py (hunk 2 above)
Comment (gitsridhar) [reply ocs_ci/deployment/ibm_deployment.py]: Completed.

File: ocs_ci/deployment/ibm.py (hunk 6 above)
Comment (gitsridhar) [reply ocs_ci/deployment/ibm.py]: This is completed. The commented out code was because this was under testing.

File: conf/ocsci/production_powervs_upi.yaml [hunk 9]
```diff
@@ -0,0 +1,37 @@
+RUN:
+  username: 'kubeadmin'
+  password_location: 'auth/kubeadmin-password'
+  log_dir: "/tmp"
+  kubeconfig_location: 'auth/kubeconfig' # relative from cluster_dir
+# bin_dir: './bin'
+  bin_dir: '/usr/local/bin'
```
[the remark this answers is not in the document]
Comment (gitsridhar) [reply conf/ocsci/production_powervs_upi.yaml:7]: I found that OCS-CI downloads and unzips x86/amd64 version of oc and kubectl into ./bin directory under ocs-ci during runtime. This is why I had to move the bin_dir to another directory.

File: conf/ocsci/production_powervs_upi.yaml [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+  force_download_client: False
+  ocs_csv_channel: "stable-4.6"
+  default_latest_tag: "latest-stable-4.6.0"
+ENV_DATA:
+  cluster_name: null.  # this parameter is passed in while invoking run-ci
+  storage_cluster_name: 'ocs-storagecluster'
+  storage_device_sets_name: "ocs-deviceset"
+  cluster_namespace: 'openshift-storage'
+  ocs_version: '4.6'
+  skip_ocp_deployment: true
```
[the remark this answers is not in the document]
Comment (gitsridhar) [reply conf/ocsci/production_powervs_upi.yaml:22]: At this point yes, but not forever. OCP install through OCS-CI will be nice.

File: conf/ocsci/production_powervs_upi.yaml [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+  kubeconfig_location: 'auth/kubeconfig' # relative from cluster_dir
+# bin_dir: './bin'
+  bin_dir: '/usr/local/bin'
+
+DEPLOYMENT:
+  local_storage: True
+  force_download_installer: False
+  force_download_client: False
+  ocs_csv_channel: "stable-4.6"
+  default_latest_tag: "latest-stable-4.6.0"
```
[the remark this answers is not in the document]
Comment (gitsridhar) [reply conf/ocsci/production_powervs_upi.yaml]: At this point, we need to work with a specific version. This will change in the future once ocs-catalogsource version stabilizes.

File: ocs_ci/deployment/ibm.py [hunk 12]
```diff
@@ -0,0 +1,524 @@
...
+
+        if not config.ENV_DATA['skip_ocp_deployment']:
+            logger.info("Destroy of OCP not implemented yet.") 
+
+
+def get_device_paths(worker_names):
```
[the remark this answers is not in the document]
Comment (gitsridhar) [reply ocs_ci/deployment/ibm.py]: get_device_paths needs to refer to config in the case of Power. It cannot go and query nodes and look for directories in /dev/disk/by-id etc..

File: ocs_ci/deployment/ibm.py (hunk 7 above)
Comment (gitsridhar) [reply ocs_ci/deployment/ibm.py:370]: I duplicated deploy_ocs due to the code related to the validation of noobaa-endpoints. We need to support noobaa so this will get cleaned up in future.

File: ocs_ci/deployment/ibm.py (hunk 8 above)
Comment (gitsridhar) [reply ocs_ci/deployment/ibm.py:533]: I want to cover the case of destroying OCS only and not OCP.

File: conf/ocsci/production_powervs_upi.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (clacroix12) [reply conf/ocsci/production_powervs_upi.yaml:7]: I'm not sure I understand how moving the bin directory helps with this issue. If the `oc` and `kubectl` binaries that are being downloaded aren't compatible with the platform, we should update how we are downloading those binaries to work for the configured platform, no?

File: ocs_ci/deployment/ibm.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (clacroix12) [reply ocs_ci/deployment/ibm.py]: `get_device_paths` is currently only used from the `setup_local_storage` function in our deployment code. I think having a function here with the same name is just going to cause issues in troubleshooting later on. As the function is only two (pretty basic) lines of code, and doesn't actually use the `worker_names` argument anyways, I think we can safely remove the function and just perform those two lines where we are calling `get_device_paths` now. What do you think?

[the remark this answers is not in the document]
Comment (clacroix12) [reply ocs_ci/deployment/ibm.py]: I see now that the issue is you're also overwriting `get_local_storage`. This is another case where there is far too much duplication. This is also a standalone function that is defined in the base deployment module. If there is logic that needs to change based on the platform type, we need to change it there in the original definition of the function. If things are getting a bit too complicated, maybe we can consider breaking that function into a class that can handle the platform specific needs. For now, I think having the additional functionality that you need behind an `if` is sufficient. After looking at the diff between the original function and the one you have in `ibm.py` I don't think it will be too complicated to adjust the original `setup_local_storage` and `get_device_paths` functions to do what you need them to do if the platform type is IBM power.

TLDR: adjust the original functions in `deployment.py` to do what you need `if platform == "powervs"`.  This will reduce the amount of duplication between `ibm.py` and `deployment.py`.

File: ocs_ci/deployment/ibm.py (hunk 7 above)
Comment (clacroix12) [reply ocs_ci/deployment/ibm.py:368]: The only thing that changed is the commented out code. If this code doesn't need to execute for IBM power you simply need to use an `if` to make sure it doesn't. We don't need to duplicate the entire function to change this behavior.

File: ocs_ci/deployment/ibm.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (gitsridhar) [reply ocs_ci/deployment/ibm.py]: ok, I misunderstood earlier comments,  initially I modified deployment.py, and then reviewers (ebenahar) wanted me to create a subclass of Deployment into ibm.py. Now I understand that I should do both with minimal duplication and it is ok to change deployment.py with conditional coding. I will do and raise another version of this pull request.

[the remark this answers is not in the document]
Comment (clacroix12) [reply ocs_ci/deployment/ibm.py]: You can absolutely (and should) create a subclass and override/add any methods you need to regarding the OCP deployment. Please check out the [azure](https://github.com/red-hat-storage/ocs-ci/blob/master/ocs_ci/deployment/azure.py) class for a good example of this.

Since the OCS deployment is (mostly) the same across platforms, this is handled in the base class. If there is a slight change in logic for a particular platform we can change it there and not have to copy the entire ocs deployment method. Same with the `setup_local_storage` and `get_device_paths` functions. Since these are standalone functions in that module and not a part of the class itself it's easier to just have them handle the logic change based on platform.

Feel free to ping me if you have any questions.

File: conf/ocsci/production_powervs_upi.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (gitsridhar) [reply conf/ocsci/production_powervs_upi.yaml:7]: OCS-CI code currently downloads client binaries from https://openshift-release-artifacts.svc.ci.openshift.org/. This has only x86 platform binaries. As far as I know, the binaries for ppc64le are installed with OCP. I could not find a separate link for OCS-CI to download it programmatically. So, one of the prereqs for OCS-CI on Power is to copy oc binary to /usr/local/bin to wherever OCS-CI is going to be installed (or if OCS-CI is installed and used right from bastion, then there is no action needed.). I will remove the commented line that points bin_dir to ./bin and keep bin_dir pointing to /usr/local/bin in our conf file for OCS-CI.

File: ocs_ci/deployment/ibm.py (hunk 5 above)
Comment (gitsridhar) [reply ocs_ci/deployment/ibm.py]: Code cleaned up. Comments removed.

File: ocs_ci/deployment/ibm.py [hunk 13]
```diff
@@ -0,0 +1,548 @@
...
+                )
+                update_ntp_compute_nodes()
+                assert ceph_health_check(
+                    namespace=self.namespace, tries=60, delay=10
+                )
+
+        # patch gp2/thin storage class as 'non-default'
+        self.patch_default_sc_to_non_default()
+
+        # Modify Noobaa endpoint auto scale values according to the cluster specs
```
[the remark this answers is not in the document]
Comment (gitsridhar) [reply ocs_ci/deployment/ibm.py]: Code cleaned up based on other comments. These commented code is no more.

File: conf/ocsci/production_powervs_upi.yaml [hunk 14]
```diff
@@ -0,0 +1,37 @@
+RUN:
```
[the remark this answers is not in the document]
Comment (gitsridhar) [reply conf/ocsci/production_powervs_upi.yaml:1]: I compared our conf file with default and cleaned up redundant entries.

File: conf/ocsci/production_powervs_upi.yaml [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+  username: 'kubeadmin'
+  password_location: 'auth/kubeadmin-password'
+  log_dir: "/tmp"
+  kubeconfig_location: 'auth/kubeconfig' # relative from cluster_dir
+# bin_dir: './bin'
```
[the remark this answers is not in the document]
Comment (gitsridhar) [reply conf/ocsci/production_powervs_upi.yaml]: Done.

File: ocs_ci/deployment/ibm.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (gitsridhar) [reply ocs_ci/deployment/ibm.py]: This is resolved now, removed every common method from ibm.py.

File: ocs_ci/deployment/ibm.py (hunk 7 above)
Comment (gitsridhar) [reply ocs_ci/deployment/ibm.py]: Done. Removed redundant code from ibm.py and moved IBM Power specific code to deployment.py with conditions.

File: conf/ocsci/production_powervs_upi.yaml [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+  force_download_installer: False
+  force_download_client: False
+  ocs_csv_channel: "stable-4.6"
+  default_latest_tag: "latest-stable-4.6.0"
+ENV_DATA:
+  cluster_name: null.  # this parameter is passed in while invoking run-ci
+  storage_cluster_name: 'ocs-storagecluster'
+  storage_device_sets_name: "ocs-deviceset"
+  cluster_namespace: 'openshift-storage'
+  ocs_version: '4.6'
```
[the remark this answers is not in the document]
Comment (gitsridhar) [reply conf/ocsci/production_powervs_upi.yaml:12]: This and next comment about specific version will be resolved after testing with command line parameter of --ocs-version.

File: conf/ocsci/production_powervs_upi.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (gitsridhar) [reply conf/ocsci/production_powervs_upi.yaml]: See above, after testing, I will resolve this comment.

File: ocs_ci/deployment/ibm.py [hunk 17]
```diff
@@ -0,0 +1,227 @@
...
+            scdevices = localvolume['storageClassDevices']
+            devicepaths = [sub['devicePaths'] for sub in scdevices]
+            alldevicepaths.append(devicepaths[0][0])
+
+        # STEP 5
+        # delete storage cluster - all of them.
+        scdelete = 'oc delete -n openshift-storage storagecluster --all --wait=true'
+        logger.info(scdelete)
+        out = run_cmd(scdelete)
+        logger.info(out)
```
Comment (clacroix12) [inline ocs_ci/deployment/ibm.py:105]: `out` should already be being logged by `run_cmd`. I don't think we need to duplicate this here. Address on all other instances of this pattern here in `ibm.py`.
